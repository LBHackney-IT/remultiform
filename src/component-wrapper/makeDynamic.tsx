import PropTypes from "prop-types";
import React from "react";
import { Subtract } from "utility-types";

import {
  DynamicComponent,
  DynamicComponentControlledProps
} from "./DynamicComponent";

/**
 * The proptypes for the component returned by {@link makeDynamic}.
 */
export type DynamicProps<
  Props extends PropsToMap,
  PropsToMap extends {},
  Value
> = Subtract<Props, PropsToMap> & DynamicComponentControlledProps<Value>;

/**
 * A map of dynamic controlled prop names to component prop names for passing to
 * {@link makeDynamic}.
 *
 * See {@link DynamicComponentControlledProps} for the type of each prop
 * and how it's used.
 */
export interface MakeDynamicPropMap<PropsToMap> {
  /**
   * The name of the prop on the wrapped component to map `value` to.
   */
  value: keyof PropsToMap & string;

  /**
   * The name of the prop on the wrapped component to map `onValueChange` to.
   */
  onValueChange: keyof PropsToMap & string;

  /**
   * The name of the prop on the wrapped component to map `disabled` to.
   */
  disabled: keyof PropsToMap & string;
}

/**
 * An adapter to pass to {@link makeDynamic} to map
 * {@link DynamicComponentControlledProps.onValueChange} to its
 * corresponding prop in the wrapped component.
 */
export interface MakeDynamicValueChangeAdapter<Value> {
  /**
   * This function should take any arguments from the event you're listening for
   * and convert it to `Value` ready to be consumed by
   * {@link DynamicComponentControlledProps.onValueChange}.
   */
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  (...args: any[]): Value;
}

const mapDynamicPropsToComponentProps = <
  Props extends PropsToMap,
  PropsToMap extends {},
  Value
>(
  dynamicProps: Readonly<
    Subtract<Props, PropsToMap> & DynamicComponentControlledProps<Value>
  >,
  { value, onValueChange, disabled }: MakeDynamicPropMap<PropsToMap>,
  valueChangeAdapter: MakeDynamicValueChangeAdapter<Value>
): Props => {
  const copyOfDynamicProps = {
    ...dynamicProps
  } as Subtract<Props, PropsToMap> &
    Partial<DynamicComponentControlledProps<Value>>;

  delete copyOfDynamicProps.value;
  delete copyOfDynamicProps.onValueChange;
  delete copyOfDynamicProps.disabled;

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const handleValueChange = (...args: any[]): void => {
    const value = valueChangeAdapter(...args);

    dynamicProps.onValueChange(value);
  };

  return {
    ...copyOfDynamicProps,
    [value]: dynamicProps.value,
    [onValueChange]: handleValueChange,
    [disabled]: dynamicProps.disabled
  } as Props;
};

const mapComponentPropTypesToDynamicPropTypes = <
  Props extends PropsToMap,
  PropsToMap extends {},
  Value
>(
  componentPropTypes: Readonly<React.WeakValidationMap<Props>> | undefined,
  { value, onValueChange, disabled }: MakeDynamicPropMap<PropsToMap>
): React.WeakValidationMap<Subtract<Props, PropsToMap> &
  DynamicComponentControlledProps<Value>> => {
  const copyOfComponentPropTypes = {
    ...componentPropTypes
  } as PropTypes.ValidationMap<
    Subtract<Props, PropsToMap> & Partial<PropsToMap>
  >;

  delete copyOfComponentPropTypes[value];
  delete copyOfComponentPropTypes[onValueChange];
  delete copyOfComponentPropTypes[disabled];

  const valuePropType = componentPropTypes
    ? ((componentPropTypes[value] as unknown) as PropTypes.Validator<Value>)
    : PropTypes.any.isRequired;

  return {
    ...copyOfComponentPropTypes,
    ...DynamicComponent.controlledPropTypes<Value>(valuePropType)
  } as React.WeakValidationMap<
    Subtract<Props, PropsToMap> & DynamicComponentControlledProps<Value>
  >;
};

/**
 * A {@link https://reactjs.org/docs/higher-order-components.html|Higher Order
 * Component (HOC)} for mapping {@link DynamicComponentControlledProps} to
 * a component's props without the hassle of creating a wrapper component.
 *
 * You will need to provide the type parameters to have this function correctly.
 *
 * @typeparam ComponentType - The type of the React component wrapped by this
 * HOC.
 *
 * @typeparam Props - The proptypes for `Component`.
 *
 * @typeparam PropNamesToMap - A union of the names of the props in `Props`
 * that need to be mapped to {@link DynamicComponentControlledProps} via
 * `propMap`.
 *
 * ```ts
 * type PropNamesToMap = "inputValue" | "onInputValueChange" | "inputDisabled";
 * ```
 *
 * **Do not use `string` or all props will be removed from the PropTypes!**
 *
 * @typeparam Value - The type of the
 * {@link DynamicComponentControlledProps.value} to be used by `Component`.
 *
 * @param Component - The React component to wrap.
 *
 * @param propMap - The map of {@link DynamicComponentControlledProps} keys
 * to `PropNamesToMap`.
 *
 * @returns A React component, ready to be used in a
 * {@link DynamicComponent}, with its
 * {@link DynamicComponentControlledProps} mapped to the `Component` as
 * specified by the `propMap`.
 */
export const makeDynamic = <
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  Props extends { [K in PropNamesToMap]?: any },
  PropNamesToMap extends string,
  Value
>(
  Component: React.ElementType<Props>,
  propMap: MakeDynamicPropMap<Pick<Props, PropNamesToMap>>,
  valueChangeAdapter: MakeDynamicValueChangeAdapter<Value>
): React.FunctionComponent<Subtract<Props, Pick<Props, PropNamesToMap>> &
  DynamicComponentControlledProps<Value>> => {
  type PropsToMap = Pick<Props, PropNamesToMap>;
  type DynamicProps = Subtract<Props, Pick<Props, PropNamesToMap>> &
    DynamicComponentControlledProps<Value>;

  const Dynamic: React.FunctionComponent<DynamicProps> = (
    props: DynamicProps
  ): JSX.Element => {
    const mappedProps = mapDynamicPropsToComponentProps<
      Props,
      PropsToMap,
      Value
    >(props, propMap, valueChangeAdapter);

    if (typeof Component === "string") {
      const IntrinsicElement = Component as keyof JSX.IntrinsicElements;

      return <IntrinsicElement {...mappedProps} />;
    } else {
      return <Component {...mappedProps} />;
    }
  };

  if (typeof Component === "string") {
    const IntrinsicElement = Component as keyof JSX.IntrinsicElements;

    Dynamic.displayName = `makeDynamic(${IntrinsicElement})`;
    Dynamic.propTypes = DynamicComponent.controlledPropTypes(
      // As this is an intrinsic element, we don't know what type `value` will
      // be at runtime.
      PropTypes.any.isRequired
    ) as React.WeakValidationMap<DynamicProps>;
  } else {
    Dynamic.displayName = `makeDynamic(${Component.displayName ||
      Component.name ||
      "unknown"})`;
    Dynamic.propTypes = mapComponentPropTypesToDynamicPropTypes<
      Props,
      PropsToMap,
      Value
    >(Component.propTypes, propMap);
  }

  return Dynamic;
};
