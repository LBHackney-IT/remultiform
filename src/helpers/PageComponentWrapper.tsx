import PropTypes from "prop-types";
import React from "react";

/**
 * A component along with the associated properties needed to render it as part
 * of a {@link Step}.
 *
 * ```ts
 * const component: PageComponent<typeof MyInput> = {
 *   key: "my-input",
 *   Component: MyInput,
 *   props: {
 *     defaultValue: "Enter something?"
 *   }
 * }
 * ```
 */
export interface PageComponent<
  ComponentType extends React.ElementType<Props>,
  Props = ComponentType extends React.ElementType<infer T> ? T : {}
> {
  /**
   * A unique identifier for this component on the page.
   */
  key: React.Key;

  /**
   * The component class, function, or tag name to render.
   */
  Component: ComponentType;

  /**
   * The props to pass to {@link PageComponent.Component}.
   */
  props: JSX.LibraryManagedAttributes<ComponentType, Props>;
}

/**
 * A wrapper for {@link PageComponent} for a {@link Step}.
 *
 * Create these using {@link wrapPageComponent}.
 */
export interface PageComponentWrapper {
  key: React.Key;

  /**
   * The function to render the component.
   *
   * @param key - The key to give the generated component. Defaults to
   * {@link PageComponentWrapper.key}.
   */
  render(key?: React.Key): JSX.Element;
}

/**
 * The proptypes of a {@link PageComponentWrapper}.
 */
export const pageComponentWrapperPropType = PropTypes.exact({
  key: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  render: PropTypes.func.isRequired
}).isRequired;

/**
 * Wrap a {@link PageComponent} in a {@link PageComponentWrapper} ready to be
 * included in a {@link Step}.
 */
// This method mostly exists so we can have strict type inference on our
// components. By wrapping the component in a generic function, the type of the
// component has to be infered in order to call the function, meaning we get
// strong types with little effort.
export const wrapPageComponent = <
  ComponentType extends React.ElementType<Props>,
  Props = ComponentType extends React.ElementType<infer T> ? T : {}
>(
  component: PageComponent<ComponentType, Props>
): PageComponentWrapper => {
  const { key, Component, props } = component;

  return {
    key,
    render(k: React.Key = key): JSX.Element {
      return <Component key={k} {...props} />;
    }
  };
};
