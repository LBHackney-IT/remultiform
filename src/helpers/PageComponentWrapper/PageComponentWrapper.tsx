import PropTypes from "prop-types";
import React from "react";

import { WrappedPageComponent } from "../../components/WrappedPageComponent";

import { NamedSchema, Schema, StoreNames, StoreValue } from "../../store/types";

import {
  DynamicPageComponentType,
  DynamicPageComponent
} from "./DynamicPageComponent";
import { StaticPageComponent } from "./StaticPageComponent";

/**
 * A wrapper for a {@link StaticPageComponent} or {@link DynamicPageComponent}
 * for use in a {@link Step}.
 *
 * Create these using {@link PageComponentWrapper.wrapStatic} and
 * {@link PageComponentWrapper.wrapDynamic}.
 */
// This mostly exists so we can have strict type inference on our components.
// By wrapping the component in a generic function, the type of the component
// has to be infered in order to call the function. Returning this non-generic
// type from that function means we can have non-generic components depend on
// it. This gives us strong typing with little effort for the user.
export class PageComponentWrapper {
  /**
   * The proptype validator for a {@link PageComponentWrapper}.
   */
  static readonly propType: PropTypes.Requireable<
    PageComponentWrapper
  > = PropTypes.exact({
    key: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
    element: PropTypes.element.isRequired
  });

  /**
   * Wrap a {@link StaticPageComponent} in a {@link PageComponentWrapper} ready
   * to be included in a {@link Step}.
   *
   * You shouldn't need to provide any of the type parameters. They should be
   * infered from the {@link StaticPageComponent} passed in.
   */
  static wrapStatic<Props>(
    component: StaticPageComponent<React.ElementType<Props>, Props>
  ): PageComponentWrapper {
    const { key, Component } = component;

    let element: JSX.Element;

    if (typeof Component === "string") {
      const {
        Component: IntrinsicElement,
        props
      } = component as StaticPageComponent<keyof JSX.IntrinsicElements, {}>;

      element = <IntrinsicElement key={key} {...props} />;
    } else {
      const {
        Component: ReactComponent,
        props
      } = component as StaticPageComponent<React.ComponentType<Props>, Props>;

      element = <ReactComponent key={key} {...props} />;
    }

    return new PageComponentWrapper(key, element);
  }

  /**
   * Wrap a {@link DynamicPageComponent} in a {@link PageComponentWrapper}
   * ready to be included in a {@link Step}.
   *
   * You shouldn't need to provide any of the type parameters. They should be
   * infered from the {@link DynamicPageComponent} passed in.
   */
  static wrapDynamic<
    Props,
    DBSchema extends NamedSchema<string, number, Schema>,
    StoreName extends StoreNames<DBSchema["schema"]>
  >(
    component: DynamicPageComponent<
      DynamicPageComponentType<
        Props,
        StoreValue<DBSchema["schema"], StoreName>
      >,
      Props,
      DBSchema,
      StoreName
    >
  ): PageComponentWrapper {
    const { key } = component;

    return new PageComponentWrapper(
      key,
      (
        <WrappedPageComponent<Props, DBSchema, StoreName>
          key={key}
          component={component}
        />
      ),
      true
    );
  }

  readonly key: React.Key;

  /**
   * The element for including in a page.
   */
  readonly element: JSX.Element;

  /**
   * Do not use this directly. Use {@link PageComponentWrapper.wrapStatic} or
   * {@link PageComponentWrapper.wrapDynamic} to create a new
   * {@link PageComponentWrapper}.
   *
   * @ignore
   */
  constructor(key: React.Key, element: JSX.Element, isDynamic = false) {
    this.key = key;

    if (isDynamic && element.type.name !== WrappedPageComponent.name) {
      throw new Error(
        `Expected element to be ${WrappedPageComponent.name} but ` +
          `received ${element.type}`
      );
    }

    this.element = element;
  }
}
