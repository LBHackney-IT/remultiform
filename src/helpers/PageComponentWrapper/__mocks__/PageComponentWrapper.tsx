import React from "react";

import { Database } from "../../../store/Database";
import {
  NamedSchema,
  Schema,
  StoreNames,
  StoreValue
} from "../../../store/types";

import { DatabaseMap } from "../DatabaseMap";
import {
  DynamicPageComponentType,
  DynamicPageComponent
} from "../DynamicPageComponent";
import { StaticPageComponent } from "../StaticPageComponent";

const { PageComponentWrapper: ActualPageComponentWrapper } = jest.requireActual(
  "../PageComponentWrapper"
);

export { ActualPageComponentWrapper };

export interface PageComponentWrapperRenderProps<
  DBSchema extends NamedSchema<string, number, Schema>,
  StoreName extends StoreNames<DBSchema["schema"]>
> {
  database?: Database<DBSchema>;
  onChange(value: StoreValue<DBSchema["schema"], StoreName>): void;
}

export class PageComponentWrapper<
  DBSchema extends NamedSchema<string, number, Schema>,
  StoreName extends StoreNames<DBSchema["schema"]>
> extends ActualPageComponentWrapper<DBSchema, StoreName> {
  static wrapStatic<Props>(
    component: StaticPageComponent<React.ElementType<Props>, Props>
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  ): PageComponentWrapper<any, string> {
    const { key, Component } = component;

    return new PageComponentWrapper(key, () => (
      <div key={key} data-testid={key}>
        Wrapped{" "}
        {typeof Component === "string"
          ? Component
          : Component.displayName || Component.name}
      </div>
    ));
  }

  static wrapDynamic<
    Props,
    DBSchema extends NamedSchema<string, number, Schema>,
    StoreName extends StoreNames<DBSchema["schema"]>
  >(
    component: DynamicPageComponent<
      DynamicPageComponentType<
        Props,
        StoreValue<DBSchema["schema"], StoreName>
      >,
      Props,
      DBSchema,
      StoreName
    >
  ): PageComponentWrapper<DBSchema, StoreName> {
    const { key, Component, databaseMap, defaultValue, emptyValue } = component;

    return new PageComponentWrapper(
      key,
      () => (
        <div key={key} data-testid={key}>
          Wrapped {Component.displayName || Component.name}
        </div>
      ),
      databaseMap,
      defaultValue,
      emptyValue
    );
  }

  constructor(
    key: string,
    render: (
      props: PageComponentWrapperRenderProps<DBSchema, StoreName>
    ) => JSX.Element,
    databaseMap?: DatabaseMap<DBSchema, StoreName>,
    defaultValue?: StoreValue<DBSchema["schema"], StoreName> | null,
    emptyValue?: "" | StoreValue<DBSchema["schema"], StoreName>
  ) {
    super(key, render, databaseMap, defaultValue, emptyValue);
  }
}
