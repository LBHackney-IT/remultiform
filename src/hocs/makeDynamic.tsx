import PropTypes from "prop-types";
import React from "react";
import { Subtract } from "utility-types";

import {
  DynamicPageComponent,
  DynamicPageComponentControlledProps
} from "../helpers/PageComponentWrapper/DynamicPageComponent";

/**
 * The proptypes for the component returned by {@link makeDynamic}.
 */
export type DynamicProps<
  Props extends PropsToMap,
  PropsToMap extends {},
  Value
> = Subtract<Props, PropsToMap> & DynamicPageComponentControlledProps<Value>;

/**
 * A map of dynamic controlled prop names to component prop names for passing to
 * {@link makeDynamic}.
 *
 * See {@link DynamicPageComponentControlledProps} for the type of each prop
 * and how it's used.
 */
export interface MakeDynamicPropMap<PropsToMap> {
  /**
   * The name of the prop on the wrapped component to map `value` to.
   */
  value: keyof PropsToMap;

  /**
   * The name of the prop on the wrapped component to map `onValueChange` to.
   */
  onValueChange: keyof PropsToMap;

  /**
   * The name of the prop on the wrapped component to map `disabled` to.
   */
  disabled: keyof PropsToMap;
}

/**
 * An adapter to pass to {@link makeDynamic} to map
 * {@link DynamicPageComponentControlledProps.onValueChange} to its
 * corresponding prop in the wrapped component.
 */
export interface MakeDynamicValueChangeAdapter<Value> {
  /**
   * This function should take any arguments from the event you're listening for
   * and convert it to `Value` ready to be consumed by
   * {@link DynamicPageComponentControlledProps.onValueChange}.
   */
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  (...args: any[]): Value;
}

const mapDynamicPropsToComponentProps = <
  Props extends PropsToMap,
  PropsToMap extends {},
  Value
>(
  dynamicProps: Readonly<
    Subtract<Props, PropsToMap> & DynamicPageComponentControlledProps<Value>
  >,
  { value, onValueChange, disabled }: MakeDynamicPropMap<PropsToMap>,
  valueChangeAdapter: MakeDynamicValueChangeAdapter<Value>
): Props => {
  const copyOfDynamicProps = {
    ...dynamicProps
  } as Subtract<Props, PropsToMap> &
    Partial<DynamicPageComponentControlledProps<Value>>;

  delete copyOfDynamicProps.value;
  delete copyOfDynamicProps.onValueChange;
  delete copyOfDynamicProps.disabled;

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const handleValueChange = (...args: any[]): void => {
    const value = valueChangeAdapter(...args);

    dynamicProps.onValueChange(value);
  };

  return {
    ...copyOfDynamicProps,
    [value]: dynamicProps.value,
    [onValueChange]: handleValueChange,
    [disabled]: dynamicProps.disabled
  } as Props;
};

const mapComponentPropTypesToDynamicPropTypes = <
  Props extends PropsToMap,
  PropsToMap extends {},
  Value
>(
  componentPropTypes: Readonly<React.WeakValidationMap<Props>> | undefined,
  { value, onValueChange, disabled }: MakeDynamicPropMap<PropsToMap>
): React.WeakValidationMap<Subtract<Props, PropsToMap> &
  DynamicPageComponentControlledProps<Value>> => {
  const copyOfComponentPropTypes = {
    ...componentPropTypes
  } as PropTypes.ValidationMap<
    Subtract<Props, PropsToMap> & Partial<PropsToMap>
  >;

  delete copyOfComponentPropTypes[value];
  delete copyOfComponentPropTypes[onValueChange];
  delete copyOfComponentPropTypes[disabled];

  const valuePropType = componentPropTypes
    ? ((componentPropTypes[value] as unknown) as PropTypes.Requireable<Value>)
    : PropTypes.any;

  return {
    ...copyOfComponentPropTypes,
    ...DynamicPageComponent.controlledPropTypes(valuePropType)
  } as React.WeakValidationMap<
    Subtract<Props, PropsToMap> & DynamicPageComponentControlledProps<Value>
  >;
};

/**
 * A {@link https://reactjs.org/docs/higher-order-components.html|Higher Order
 * Component (HOC)} for mapping {@link DynamicPageComponentControlledProps} to
 * a component's props without the hassle of creating a wrapper component.
 *
 * You will need to provide the type parameters to have this function correctly.
 *
 * @typeparam ComponentType - The type of the React component wrapped by this
 * HOC.
 *
 * @typeparam Props - The proptypes for `Component`.
 *
 * @typeparam PropNamesToMap - A union of the names of the props in `Props`
 * that need to be mapped to {@link DynamicPageComponentControlledProps} via
 * `propMap`.
 *
 * ```ts
 * type PropNamesToMap = "inputValue" | "onInputValueChange" | "inputDisabled";
 * ```
 *
 * **Do not use `string` or all props will be removed from the PropTypes!**
 *
 * @typeparam Value - The type of the
 * {@link DynamicPageComponentControlledProps.value} to be used by `Component`.
 *
 * @param Component - The React component to wrap.
 *
 * @param propMap - The map of {@link DynamicPageComponentControlledProps} keys
 * to `PropNamesToMap`.
 *
 * @returns A React component, ready to be used in a
 * {@link DynamicPageComponent}, with its
 * {@link DynamicPageComponentControlledProps} mapped to the `Component` as
 * specified by the `propMap`.
 */
export const makeDynamic = <
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  Props extends { [K in PropNamesToMap]?: any },
  PropNamesToMap extends string,
  Value
>(
  Component: React.ElementType<Props>,
  propMap: MakeDynamicPropMap<Pick<Props, PropNamesToMap>>,
  valueChangeAdapter: MakeDynamicValueChangeAdapter<Value>
): React.FunctionComponent<Subtract<Props, Pick<Props, PropNamesToMap>> &
  DynamicPageComponentControlledProps<Value>> => {
  type PropsToMap = Pick<Props, PropNamesToMap>;
  type DynamicProps = Subtract<Props, Pick<Props, PropNamesToMap>> &
    DynamicPageComponentControlledProps<Value>;

  const Dynamic: React.FunctionComponent<DynamicProps> = (
    props: DynamicProps
  ): JSX.Element => {
    const mappedProps = mapDynamicPropsToComponentProps<
      Props,
      PropsToMap,
      Value
    >(props, propMap, valueChangeAdapter);

    if (typeof Component === "string") {
      const IntrinsicElement = Component as keyof JSX.IntrinsicElements;
      // Infering the type of the React props acceptable for any possible
      // `JSX.IntrinsicElements` is Hard. The type of this never leaves this
      // scope, so we cheat, given that the author is confident about the types
      // actually matching, whatever TypeScript says.
      const intrinsicElementProps = mappedProps as {};

      return <IntrinsicElement {...intrinsicElementProps} />;
    } else {
      return <Component {...mappedProps} />;
    }
  };

  if (typeof Component === "string") {
    const IntrinsicElement = Component as keyof JSX.IntrinsicElements;

    Dynamic.displayName = `makeDynamic(${IntrinsicElement})`;

    // We don't know the type of `Value` at runtime to use for the proptypes,
    // so we allow `any`.
    Dynamic.propTypes = DynamicPageComponent.controlledPropTypes<Value>(
      PropTypes.any
    ) as React.WeakValidationMap<DynamicProps>;
  } else {
    Dynamic.displayName = `makeDynamic(${Component.displayName ||
      Component.name ||
      "unknown"})`;
    Dynamic.propTypes = mapComponentPropTypesToDynamicPropTypes<
      Props,
      PropsToMap,
      Value
    >(Component.propTypes, propMap);
  }

  return Dynamic;
};
