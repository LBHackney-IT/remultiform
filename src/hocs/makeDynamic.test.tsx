import { nullValuesAsUndefined } from "null-as-undefined";
import PropTypes from "prop-types";
import React from "react";
import { create } from "react-test-renderer";

import { spyOnConsoleError } from "../__tests__/helpers/spies";

import { DynamicPageComponent } from "../helpers/PageComponentWrapper/DynamicPageComponent";

import { makeDynamic } from "./makeDynamic";

it("renders correctly for function components", () => {
  interface TestProps {
    className: string;
    type: "text";

    inputValue?: string | null;
    onInputValueChange(value: React.ChangeEvent<HTMLInputElement>): void;
    inputDisabled: boolean;
  }

  const TestFunctionComponent: React.FunctionComponent<TestProps> = (
    props: TestProps
  ): JSX.Element => {
    const {
      className,
      type,
      inputValue,
      onInputValueChange,
      inputDisabled
    } = nullValuesAsUndefined(props);

    return (
      <input
        className={className}
        type={type}
        value={inputValue}
        onChange={onInputValueChange}
        disabled={inputDisabled}
      />
    );
  };

  const Component = makeDynamic<
    TestProps,
    "inputValue" | "onInputValueChange" | "inputDisabled",
    string
  >(
    TestFunctionComponent,
    {
      value: "inputValue",
      onValueChange: "onInputValueChange",
      disabled: "inputDisabled"
    },
    (event: React.ChangeEvent<HTMLInputElement>) => event.target.value
  );

  const component = create(
    <Component
      className="test-class"
      type="text"
      value="test value"
      onValueChange={jest.fn()}
      disabled={true}
    />
  );

  expect(component).toMatchSnapshot();
});

it("renders correctly for class components", () => {
  interface TestProps {
    className: string;
    type: "text";

    inputValue?: string | null;
    onInputValueChange(value: React.ChangeEvent<HTMLInputElement>): void;
    inputDisabled: boolean;
  }

  class TestClassComponent extends React.Component<TestProps, never, never> {
    render(): JSX.Element {
      const {
        className,
        type,
        inputValue,
        onInputValueChange,
        inputDisabled
      } = nullValuesAsUndefined(this.props);

      return (
        <input
          className={className}
          type={type}
          value={inputValue}
          onChange={onInputValueChange}
          disabled={inputDisabled}
        />
      );
    }
  }

  const Component = makeDynamic<
    TestProps,
    "inputValue" | "onInputValueChange" | "inputDisabled",
    string
  >(
    TestClassComponent,
    {
      value: "inputValue",
      onValueChange: "onInputValueChange",
      disabled: "inputDisabled"
    },
    (event: React.ChangeEvent<HTMLInputElement>) => event.target.value
  );

  const component = create(
    <Component
      className="test-class"
      type="text"
      value="test value"
      onValueChange={jest.fn()}
      disabled={true}
    />
  );

  expect(component).toMatchSnapshot();
});

it("creates a component with a sensible display name when the original component is an intrinsic element", () => {
  const Component = makeDynamic<
    React.InputHTMLAttributes<HTMLInputElement>,
    "value" | "onChange" | "disabled",
    string
  >(
    "input",
    {
      value: "value",
      onValueChange: "onChange",
      disabled: "disabled"
    },
    (event: React.ChangeEvent<HTMLInputElement>) => event.target.value
  );

  expect(Component.displayName).toEqual("makeDynamic(input)");
});

it("creates a component with a sensible display name when the original component is a function component", () => {
  interface TestProps {
    className: string;
    type: "text";

    inputValue?: string | null;
    onInputValueChange(value: React.ChangeEvent<HTMLInputElement>): void;
    inputDisabled: boolean;
  }

  const TestFunctionComponent: React.FunctionComponent<TestProps> = (
    props: TestProps
  ): JSX.Element => {
    const {
      className,
      type,
      inputValue,
      onInputValueChange,
      inputDisabled
    } = nullValuesAsUndefined(props);

    return (
      <input
        className={className}
        type={type}
        value={inputValue}
        onChange={onInputValueChange}
        disabled={inputDisabled}
      />
    );
  };

  const Component = makeDynamic<
    TestProps,
    "inputValue" | "onInputValueChange" | "inputDisabled",
    string
  >(
    TestFunctionComponent,
    {
      value: "inputValue",
      onValueChange: "onInputValueChange",
      disabled: "inputDisabled"
    },
    (event: React.ChangeEvent<HTMLInputElement>) => event.target.value
  );

  expect(Component.displayName).toEqual("makeDynamic(TestFunctionComponent)");
});

it("creates a component with a sensible display name when the original component is a class component", () => {
  interface TestProps {
    className: string;
    type: "text";

    inputValue?: string | null;
    onInputValueChange(value: React.ChangeEvent<HTMLInputElement>): void;
    inputDisabled: boolean;
  }

  class TestClassComponent extends React.Component<TestProps, never, never> {
    render(): JSX.Element {
      const {
        className,
        type,
        inputValue,
        onInputValueChange,
        inputDisabled
      } = nullValuesAsUndefined(this.props);

      return (
        <input
          className={className}
          type={type}
          value={inputValue}
          onChange={onInputValueChange}
          disabled={inputDisabled}
        />
      );
    }
  }

  const Component = makeDynamic<
    TestProps,
    "inputValue" | "onInputValueChange" | "inputDisabled",
    string
  >(
    TestClassComponent,
    {
      value: "inputValue",
      onValueChange: "onInputValueChange",
      disabled: "inputDisabled"
    },
    (event: React.ChangeEvent<HTMLInputElement>) => event.target.value
  );

  expect(Component.displayName).toEqual("makeDynamic(TestClassComponent)");
});

it("creates a component with the correct proptypes when the original component is an intrinsic element", () => {
  const Component = makeDynamic<
    React.InputHTMLAttributes<HTMLInputElement>,
    "value" | "onChange" | "disabled",
    string
  >(
    "input",
    {
      value: "value",
      onValueChange: "onChange",
      disabled: "disabled"
    },
    (event: React.ChangeEvent<HTMLInputElement>) => event.target.value
  );

  expect(Component.propTypes).toEqual(
    DynamicPageComponent.controlledPropTypes(PropTypes.any)
  );
});

it("creates a component with the correct proptypes when the original component has no proptypes", () => {
  interface TestProps {
    className: string;
    type: "text";

    inputValue?: string | null;
    onInputValueChange(value: React.ChangeEvent<HTMLInputElement>): void;
    inputDisabled: boolean;
  }

  const TestFunctionComponent: React.FunctionComponent<TestProps> = (
    props: TestProps
  ): JSX.Element => {
    const {
      className,
      type,
      inputValue,
      onInputValueChange,
      inputDisabled
    } = nullValuesAsUndefined(props);

    return (
      <input
        className={className}
        type={type}
        value={inputValue}
        onChange={onInputValueChange}
        disabled={inputDisabled}
      />
    );
  };

  const Component = makeDynamic<
    TestProps,
    "inputValue" | "onInputValueChange" | "inputDisabled",
    string
  >(
    TestFunctionComponent,
    {
      value: "inputValue",
      onValueChange: "onInputValueChange",
      disabled: "inputDisabled"
    },
    (event: React.ChangeEvent<HTMLInputElement>) => event.target.value
  );

  expect(Component.propTypes).toEqual(
    DynamicPageComponent.controlledPropTypes(PropTypes.any)
  );
});

it("creates a component with the correct proptypes when the original component has proptypes", () => {
  interface TestProps {
    className: string;
    type: "text";

    inputValue?: string | null;
    onInputValueChange(value: React.ChangeEvent<HTMLInputElement>): void;
    inputDisabled: boolean;
  }

  const TestFunctionComponent: React.FunctionComponent<TestProps> = (
    props: TestProps
  ): JSX.Element => {
    const {
      className,
      type,
      inputValue,
      onInputValueChange,
      inputDisabled
    } = nullValuesAsUndefined(props);

    return (
      <input
        className={className}
        type={type}
        value={inputValue}
        onChange={onInputValueChange}
        disabled={inputDisabled}
      />
    );
  };

  TestFunctionComponent.propTypes = {
    className: PropTypes.string.isRequired,
    type: PropTypes.oneOf<"text">(["text"]).isRequired,
    inputValue: PropTypes.string,
    onInputValueChange: PropTypes.func.isRequired,
    inputDisabled: PropTypes.bool.isRequired
  };

  const Component = makeDynamic<
    TestProps,
    "inputValue" | "onInputValueChange" | "inputDisabled",
    string
  >(
    TestFunctionComponent,
    {
      value: "inputValue",
      onValueChange: "onInputValueChange",
      disabled: "inputDisabled"
    },
    (event: React.ChangeEvent<HTMLInputElement>) => event.target.value
  );

  const consoleErrorSpy = spyOnConsoleError();

  PropTypes.checkPropTypes(
    Component.propTypes,
    {
      className: "class-name",
      type: "text",
      value: "value",
      onValueChange: () => {},
      disabled: true
    },
    "prop",
    Component.displayName || "unknown"
  );

  expect(consoleErrorSpy).not.toHaveBeenCalled();

  PropTypes.checkPropTypes(
    Component.propTypes,
    {
      className: 0,
      type: "not-text",
      value: 1,
      onValueChange: "on-change",
      disabled: "disabled"
    },
    "prop",
    Component.displayName || "unknown"
  );

  expect(consoleErrorSpy.mock.calls).toMatchSnapshot();
});
