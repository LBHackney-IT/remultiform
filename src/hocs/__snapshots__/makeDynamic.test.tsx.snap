// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates a component with the correct proptypes when the original component has proptypes 1`] = `
Array [
  Array [
    "Warning: Failed prop type: Invalid prop \`className\` of type \`number\` supplied to \`makeDynamic(TestFunctionComponent)\`, expected \`string\`.",
  ],
  Array [
    "Warning: Failed prop type: Invalid prop \`type\` of value \`not-text\` supplied to \`makeDynamic(TestFunctionComponent)\`, expected one of [\\"text\\"].",
  ],
  Array [
    "Warning: Failed prop type: Invalid prop \`value\` of type \`number\` supplied to \`makeDynamic(TestFunctionComponent)\`, expected \`string\`.",
  ],
  Array [
    "Warning: Failed prop type: Invalid prop \`onValueChange\` of type \`string\` supplied to \`makeDynamic(TestFunctionComponent)\`, expected \`function\`.",
  ],
  Array [
    "Warning: Failed prop type: Invalid prop \`disabled\` of type \`string\` supplied to \`makeDynamic(TestFunctionComponent)\`, expected \`boolean\`.",
  ],
]
`;

exports[`renders correctly for class components 1`] = `
<input
  className="test-class"
  disabled={true}
  onChange={[Function]}
  type="text"
  value="test value"
/>
`;

exports[`renders correctly for function components 1`] = `
<input
  className="test-class"
  disabled={true}
  onChange={[Function]}
  type="text"
  value="test value"
/>
`;
